### PR TITLE
feat: stubs variant route support with rewrite 

### PR DESCRIPTION
**Pull request by kendallstrautman**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-ksvariants-middleware-rewrite-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway?variants=config:hcp) 🔎
- [Asana task](https://app.asana.com/0/home/1199634971443477/1204343104798012) 🎟️

## 🗒️ What

This PR stubs in the variant route support and middleware rewrites for the query param, `variants`.

❗ _Merging into an assembly_ 

See the [routing section of the RFC ](https://docs.google.com/document/d/1bDQCGDgjJouBAOfc8K5zkjeSjOFvg8gC7wXPeBasf8E/edit#heading=h.mi4ha5ru0ren)for full context

## 🧪 Testing

- Visit a tutorial with[ variants defined](https://dev-portal-git-ksvariants-middleware-rewrite-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway?variants=config:hcp)
- Validate that the variant shows up in the pink box
- Validate that plain tutorial [paths without variants](https://dev-portal-git-ksvariants-middleware-rewrite-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-secure-agents) render as normal

## 💭 Anything else?

Note the bundle analysis check won't pass because the assembly branch doesn't have a PR yet, this is the first change.

### CODE REVIEW

File: src/views/tutorial-view/server.ts [hunk 1]
```diff
@@ -66,6 +69,23 @@ export async function getTutorialPageProps(
...
+	let variant = undefined
+
+	// @TODO, the variant data will be passed from the API, check for that in the
+	// tutorial data
+	if (variantSlug) {
+		// slugs in the query params are formatted like ?variants="slug:optionSlug"
+		const VARIANT_SLUG_SPLIT_CHAR = ':'
+		const [slug, optionSlug] = variantSlug.split(VARIANT_SLUG_SPLIT_CHAR)
+
+		// @TODO, expand this to pass the active variant option and all variant option data
```
Comment (kendallstrautman) [inline src/views/tutorial-view/server.ts:82]: This will get hooked up to API data.

File: src/middleware.ts [hunk 2]
```diff
@@ -122,6 +122,27 @@ export async function middleware(req: NextRequest, ev: NextFetchEvent) {
 	// 	}
 	// }
 
+	/**
+	 * Detect the variants query param and rewrite to the correct path.
+	 *
+	 * Request path: /{product}/tutorials/{collection}/{tutorial}?variants={slug:optionSlug}
```
Comment (kendallstrautman) [inline src/middleware.ts:128]: I chose to go plural here in anticipation of multi-variant world. Open to make this `variant` though if folks have a strong opinion

File: src/views/tutorial-view/index.tsx [hunk 3]
```diff
@@ -266,6 +267,12 @@ function TutorialView({
 							}}
 							tutorialId={id}
 						/>
+						{metadata.variant ? (
+							<div style={{ border: '1px solid magenta' }}>
+								<h2>Variant: {metadata.variant.slug}</h2>
+								<p>Option: {metadata.variant.optionSlug}</p>
+							</div>
+						) : null}
 						<span data-ref-id={progressRefsId} ref={progressRefs.startRef} />
```
Comment (kendallstrautman) [inline src/views/tutorial-view/index.tsx:276]: This is a stub that basically passes the data from the query param

File: src/views/tutorial-view/index.tsx [hunk 4]
```diff
@@ -266,6 +267,12 @@ function TutorialView({
 							}}
 							tutorialId={id}
 						/>
+						{metadata.variant ? (
+							<div style={{ border: '1px solid magenta' }}>
+								<h2>Variant: {metadata.variant.slug}</h2>
+								<p>Option: {metadata.variant.optionSlug}</p>
+							</div>
+						) : null}
```
Comment (kendallstrautman) [inline src/views/tutorial-view/index.tsx:275]: This is a stub that basically passes data from the query param. We will get real data from the API with the tutorial data.

**Review by zchsh — APPROVED** [review]
✅ I dig this! Followed the testing steps, all looks good. Magenta borders 4 lyfe 💅

**Pull request merged**